src/KDSoapClient/KDSoapMessage.cpp

```diff
@@ -124,12 +124,29 @@ bool KDSoapMessage::isFault() const
 
 QString KDSoapMessage::faultAsString() const
 {
-    // This better be on a single line, since it's used by server-side logging too
-    const QString actor = childValues().child(QLatin1String("faultactor")).value().toString();
-    return QObject::tr("Fault code %1: %2%3")
-           .arg(childValues().child(QLatin1String("faultcode")).value().toString())
-           .arg(childValues().child(QLatin1String("faultstring")).value().toString())
-           .arg(actor.isEmpty() ? QString() : QString::fromLatin1(" (%1)").arg(actor));
+    if(name() == QLatin1String("Fault") && namespaceUri() == QLatin1String("http://www.w3.org/2003/05/soap-envelope"))
+    {
+        QString faultCodeStr;
+        KDSoapValue faultCode = childValues().child(QLatin1String("Code"));
+        while(!faultCode.isNull()) {
```

Comment:
This while loop confused me for a while (haha, nice pun).

I suppose Subcode elements can't be nested, this was just to factorize code between Code and Subcode handling? I think it makes the code harder to read/understand.
Can you unroll the loop, i.e. handle Code and then handle Subcode, sequentially? This will duplicate only one line of code, the child("Value") line.

And it will make it possible to format the subcode stuff differently if we want to; I'm not sure it looks great right now when having a subcode (did you test it?)

Reply:
I believe Subcode can be nested. I got that from: https://www.w3.org/TR/soap12-part1/#faultsubcodeelement

Would it help if I added a comment that explains the loop?

Yes, I did test it. For example:
> Fault s:Sender ter:NotAuthorized: Sender not Authorized. Invalid username or password!